### PR TITLE
Unify Digest and Blake2bHash

### LINKED ISSUE #2008: Unify `Digest` and `Blake2bHash` Types
The `Digest` and `Blake2bHash` types duplicate functionality and should be unified.

### PR DESCRIPTION
**Pull request by dwerner**
Resolves #2008 

This PR unifies the Digest and Blake2bHash.

### CODE REVIEW

File: types/Cargo.toml [hunk 1]
```diff
@@ -62,8 +64,17 @@ std = [
...
+    "hashing/std",
+]
+no-std = [
+    "base16/alloc",
+    "hex/alloc",
+    "serde/alloc",
+    "serde_json/alloc",
+    "serde_bytes/alloc",
+    "displaydoc",
+    "hashing/alloc"
```
Comment (xcthulhu) [inline types/Cargo.toml]: @Fraser999 We ended up calling the `no-std` flag for this hashing library `alloc`, since we saw the other libraries are doing that.

Not for this PR, but maybe we follow suite and rename `no-std` -> `alloc` here for the 1.4.0 release?

File: hashing/src/lib.rs [hunk 2]
```diff
@@ -1,174 +1,165 @@
...
+        Digest(result)
+    }
+
+    /// Returns a reference to the underlying value.
+    pub fn inner(&self) -> &[u8; Digest::LENGTH] {
+        &self.0
+    }
+
+    /// Returns the underlying value.
+    pub fn value(&self) -> [u8; Digest::LENGTH] {
```
Comment (xcthulhu) [inline hashing/src/lib.rs]:
```suggestion
    pub fn value(self) -> [u8; Digest::LENGTH] {
```

Comment (xcthulhu) [reply hashing/src/lib.rs]: @dwerner when we were investigating this together this appeared to not break anything...

File: hashing/src/lib.rs [hunk 3]
```diff
@@ -1,174 +1,165 @@
...
         self.0
     }
 
-    /// Returns a copy of the wrapped `u8` array as a `Vec`
+    /// Converts the underlying hash digest array to a `Vec`
+    #[cfg(feature = "std")]
     pub fn into_vec(self) -> Vec<u8> {
         self.0.to_vec()
     }
```
Comment (xcthulhu) [inline hashing/src/lib.rs]: This is only used in 3 places, I say we just inline it...

File: client/lib/lib.rs [hunk 4]
```diff
@@ -1160,9 +1160,8 @@ impl<'a> TryInto<GlobalStateIdentifier> for GlobalStateStrParams<'a> {
     fn try_into(self) -> Result<GlobalStateIdentifier> {
         let hash = Digest::from_hex(self.hash_value).map_err(|error| Error::CryptoError {
```
Comment (afck) [inline client/lib/lib.rs:1164]: Would it be better to have a separate `HashError` variant now, instead of wrapping it in a `crypto::Error` everywhere? (Not sure; I'm not very familiar with the client code.)

(Also, I wouldn't repeat the word `Error` in the error variant names: just `Error::Crypto`.)

File: execution_engine/src/core/runtime_context/tests.rs [hunk 5]
```diff
@@ -228,6 +228,7 @@ fn account_key_not_writeable() {
 
 #[test]
 fn account_key_readable_valid() {
+    println!("evil shit");
```
Comment (afck) [inline execution_engine/src/core/runtime_context/tests.rs]: :supervillain:

File: hashing/src/lib.rs [hunk 6]
```diff
@@ -1,174 +1,165 @@
-//! Cryptographic hash type and function.
+#![cfg_attr(not(feature = "std"), no_std)]
 
-// TODO - remove once schemars stops causing warning.
-#![allow(clippy::field_reassign_with_default)]
+/// The number of bytes in a Blake2b hash
```
Comment (afck) [inline hashing/src/lib.rs]: That probably belongs to line 42?
And maybe

```rust
/// The number of bytes in a digest
```

would be more accurate: blake2b allows different sizes; 32 bytes is our choice.

File: hashing/src/lib.rs [hunk 7]
```diff
@@ -1,174 +1,165 @@
...
+    /// Hashes a pair of byte slices into a single [`Digest`]
+    pub fn hash_pair<T: AsRef<[u8]>, U: AsRef<[u8]>>(data1: T, data2: U) -> Digest {
+        let mut result = [0; Digest::LENGTH];
+        let mut hasher = VarBlake2b::new(Digest::LENGTH).unwrap();
+        hasher.update(data1);
+        hasher.update(data2);
+        hasher.finalize_variable(|slice| {
+            result.copy_from_slice(slice);
+        });
+        Digest(result)
```
Comment (afck) [inline hashing/src/lib.rs:76]: ~Is `hash_pair(&[1, 2, 3], &[4]) == hash_pair(&[1], &[2, 3, 4])`? Is that intentional?~

It looks like we're only using this in places where both inputs have a fixed size. (Also, in those places we can't change anymore how we compute the hash.)

I'd still give the method a different name, so nobody uses it assuming that it gives different results for different variable-length pairs.

Maybe `hash_concatenated`?

File: hashing/src/lib.rs [hunk 8]
```diff
@@ -1,174 +1,165 @@
...
-    }
-}
-
-impl From<[u8; Digest::LENGTH]> for Digest {
-    fn from(inner: [u8; Digest::LENGTH]) -> Self {
-        Digest(inner)
+#[cfg(feature = "std")]
+impl std::fmt::LowerHex for Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let hex_string = base16::encode_lower(self.inner());
```
Comment (afck) [inline hashing/src/lib.rs:118]: If we just iterate over `self.inner()` and write each byte we could avoid the string allocation.

File: hashing/src/lib.rs [hunk 9]
```diff
@@ -1,174 +1,165 @@
...
     }
 }
 
-impl Debug for Digest {
-    fn fmt(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
-        write!(formatter, "{}", HexFmt(&self.0))
+#[cfg(feature = "std")]
+impl std::fmt::Display for Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{:10}", base16::encode_lower(&self.0))
```
Comment (afck) [inline hashing/src/lib.rs:142]: …and here we wouldn't even have to encode all the bytes, since we're printing only five of them.

File: node/src/components/chainspec_loader.rs [hunk 10]
```diff
@@ -165,6 +164,7 @@ pub(crate) struct ChainspecLoader {
     /// If `Some`, we're finished loading and committing the chainspec.
     reactor_exit: Option<ReactorExit>,
     /// The initial state root hash for this session.
+    #[data_size(skip)]
```
Comment (afck) [inline node/src/components/chainspec_loader.rs]: Why not implement `DataSize` for `Digest`?

File: node/src/types/deploy.rs [hunk 11]
```diff
@@ -69,11 +63,11 @@ static DEPLOY: Lazy<Deploy> = Lazy::new(|| {
         ttl: TimeDiff::from(3_600_000),
         gas_price: 1,
         body_hash,
-        dependencies: vec![DeployHash::new(Digest::from([1u8; Digest::LENGTH]))],
+        dependencies: vec![DeployHash::new(Digest::hash([1u8; Digest::LENGTH]))],
```
Comment (afck) [inline node/src/types/deploy.rs:70]: `Digest::from` didn't hash the input, did it?

File: execution_engine_testing/cargo_casper/tests/integration_tests.rs [hunk 12]
```diff
@@ -73,6 +73,8 @@ fn run_tool_and_resulting_tests() {
         .arg(nightly_version.trim())
         .arg("cargo")
         .arg("test")
+        .arg("--")
+        .arg("--nocapture")
```
Comment (afck) [inline execution_engine_testing/cargo_casper/tests/integration_tests.rs]: Not sure how spammy that output is; do we always want that?

File: node/src/components/small_network/tests.rs [hunk 13]
```diff
@@ -350,6 +350,7 @@ async fn run_two_node_network_five_times() {
 /// Sanity check that we can bind to a real network.
 ///
 /// Very unlikely to ever fail on a real machine.
+#[cfg(not(target_os = "macos"))]
```
Comment (afck) [inline node/src/components/small_network/tests.rs:352]: Is this something that needs to be ticketed? Otherwise please add a comment why it doesn't work on MacOS.

File: node/src/components/small_network/tests.rs [hunk 14]
```diff
@@ -350,6 +350,7 @@ async fn run_two_node_network_five_times() {
...
 #[tokio::test]
 async fn bind_to_real_network_interface() {
     init_logging();
```
Comment (afck) [inline node/src/components/small_network/tests.rs:355]: Please also put `#[cfg(not(target_os = "macos"))]` on `use datalink`.

File: client/lib/validation.rs [hunk 15]
```diff
@@ -110,14 +109,8 @@ pub(crate) fn validate_get_era_info_response(
...
 
-            core::validate_query_proof(
-                &state_root_hash.to_owned().into(),
-                &proofs,
-                &key,
-                path,
-                &proof_value,
-            )
-            .map_err(Into::into)
+            core::validate_query_proof(&state_root_hash, &proofs, &key, path, &proof_value)
```
Comment (nilz3ro) [inline client/lib/validation.rs:113]: Why did you change the whitespace here?

File: client/lib/validation.rs [hunk 16]
```diff
@@ -173,14 +166,7 @@ pub(crate) fn validate_query_response(
...
 
-    core::validate_query_proof(
-        &state_root_hash.to_owned().into(),
-        &proofs,
-        key,
-        path,
-        proof_value,
-    )
-    .map_err(Into::into)
+    core::validate_query_proof(state_root_hash, &proofs, key, path, proof_value).map_err(Into::into)
```
Comment (nilz3ro) [inline client/lib/validation.rs:170]: ^

File: client/lib/validation.rs [hunk 17]
```diff
@@ -234,14 +220,8 @@ pub(crate) fn validate_query_global_state(
...
 
-    core::validate_query_proof(
-        &state_root_hash.to_owned().into(),
-        &proofs,
-        key,
-        path,
-        proof_value,
-    )
-    .map_err(Into::into)
+    core::validate_query_proof(&state_root_hash, &proofs, key, path, proof_value)
```
Comment (nilz3ro) [inline client/lib/validation.rs:224]: ^

File: client/lib/validation.rs [hunk 18]
```diff
@@ -280,13 +260,8 @@ pub(crate) fn validate_get_balance_response(
...
     };
 
-    core::validate_balance_proof(
-        &state_root_hash.to_owned().into(),
-        &balance_proof,
-        *key,
-        &balance,
-    )
-    .map_err(Into::into)
+    core::validate_balance_proof(state_root_hash, &balance_proof, *key, &balance)
```
Comment (nilz3ro) [inline client/lib/validation.rs:264]: ^

File: execution_engine/src/core/runtime_context/tests.rs (hunk 5 above)
Comment (nilz3ro) [reply execution_engine/src/core/runtime_context/tests.rs]: We will want to remove any curse words from our public source code

**Review by nilz3ro — COMMENTED** [review]
I marked a few places where I noticed white space changes. You can ignore these comments if it was intentional, I just wanted to make sure we don't have conflicting editor config / cargo fmt config, otherwise future PRs could be polluted with unnecessary white space changes.

File: node/src/components/small_network/tests.rs (hunk 14 above)
Comment (afck) [reply node/src/components/small_network/tests.rs:355]: Or better: just qualify `pnet::datalink` inside this test and remove the import.

File: node/src/types/deploy.rs (hunk 11 above)
Comment (dwerner) [reply node/src/types/deploy.rs:70]: No, it didn't. I think the hashed version is more appropriate in the deploy example above.

File: node/src/components/small_network/tests.rs (hunk 13 above)
Comment (dwerner) [reply node/src/components/small_network/tests.rs:352]: I believe @xcthulhu was having port binding issues on while developing on osx.

File: execution_engine_testing/cargo_casper/tests/integration_tests.rs (hunk 12 above)
Comment (dwerner) [reply execution_engine_testing/cargo_casper/tests/integration_tests.rs]: Nope, thanks.

File: client/lib/validation.rs (hunk 16 above)
Comment (dwerner) [reply client/lib/validation.rs:170]: This was `rustfmt` not an intentional change.

File: client/lib/validation.rs (hunk 17 above)
Comment (dwerner) [reply client/lib/validation.rs:224]: rustfmt

File: client/lib/validation.rs (hunk 18 above)
Comment (dwerner) [reply client/lib/validation.rs:264]: rustfmt

File: types/Cargo.toml [hunk 19]
```diff
@@ -30,14 +33,15 @@ once_cell = { version = "1.5.2", default-features = false }
 proptest = { version = "1.0.0", optional = true }
 rand = { version = "0.8.3", default-features = false, features = ["small_rng"] }
 schemars = { version = "0.8.0", features = ["preserve_order"], optional = true }
-serde = { version = "1", default-features = false, features = ["derive"] }
+serde = { version = "1", default-features = false, features = ["derive"], optional=true}
```
Comment (mpapierski) [inline types/Cargo.toml]:
```suggestion
serde = { version = "1", default-features = false, features = ["derive"], optional = true}
```

**Comment by dwerner** [comment]
Closing in favor of https://github.com/casper-network/casper-node/pull/2057
